crates/fixedpointmath/src/math.rs

```
@@ -0,0 +1,450 @@
+        }
+        let sign = self.sign().flip_if(other.sign() != divisor.sign());
+        let abs = U256::try_from(
+            self.raw()
+                .unsigned_abs()
+                .full_mul(other.raw().unsigned_abs())
+                .div(divisor.raw().unsigned_abs()),
+        )
+        .map_err(|_| eyre!("FixedPoint operation overflowed: {self} * {other} / {divisor}"))
+        .unwrap();
+        Self::from_sign_and_abs(sign, abs).unwrap()
+    }
+
+    pub fn mul_div_up(self, other: Self, divisor: Self) -> Self {
+        if divisor.is_zero() {
+            panic!("Cannot divide by zero.");
+        }
+        let sign = self.sign().flip_if(other.sign() != divisor.sign());
+        let (abs_u512, rem) = self
+            .raw()
+            .unsigned_abs()
+            .full_mul(other.raw().unsigned_abs())
+            .div_mod(divisor.raw().unsigned_abs().into());
+        let abs = U256::try_from(abs_u512)
+            .map_err(|_| eyre!("FixedPoint operation overflowed: {self} * {other} / {divisor}"))
+            .unwrap();
+        Self::from_sign_and_abs(sign, abs).unwrap()
+            + Self::from_sign_and_abs(sign, U256::from(!rem.is_zero() as u8)).unwrap()
+    }
+
+    pub fn mul_down(self, other: Self) -> Self {
+        self.mul_div_down(other, self.one())
+    }
+
+    pub fn mul_up(self, other: Self) -> Self {
+        self.mul_div_up(other, self.one())
+    }
+
+    pub fn div_down(self, other: Self) -> Self {
+        self.mul_div_down(self.one(), other)
+    }
+
+    pub fn div_up(self, other: Self) -> Self {
+        self.mul_div_up(self.one(), other)
+    }
+
+    pub fn pow(self, y: Self) -> Result<Self> {
+        let one = self.one();
+
+        // If the exponent is negative, return 1 / x^abs(y).
+        if y.is_negative() {
+            let abs_result = self.pow(y.abs())?;
+
+            // FIXME: What's the correct return value here?
+            if abs_result.is_zero() {
+                return Ok(Self::zero());
```

I would return an error here, since we're dividing by zero and that should be `nan`

---

Sounds good. I'm going to remove the check for `0` then for now and let the
error from the `div_down` below propagate until we decide a diff error or return
val here.